File: contains-duplicate/Solution217.java
// https://leetcode.com/problems/contains-duplicate/

import java.util.Arrays;
import java.util.HashSet;

class Solution217 {

    // T: O(n), S: O(n)
    public boolean containsDuplicate(int[] nums) {
        var seen = new HashSet<Integer>();
        for (int n : nums) {
            if (seen.contains(n)) {
                return true;
            }
            seen.add(n);
        }
        return false;
    }

    // T: O(n log n), S: O(1)
    public boolean containsDuplicateSort(int[] nums) {
        Arrays.sort(nums);
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] == nums[i - 1]) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        var sol = new Solution217();
        var nums1 = new int[] { 1, 2, 3, 1 };
        var nums2 = new int[] { 1, 2, 3, 4 };
        System.out.println(sol.containsDuplicate(nums1)); // true
        System.out.println(sol.containsDuplicateSort(nums1)); // true
        System.out.println(sol.containsDuplicate(nums2)); // false
        System.out.println(sol.containsDuplicateSort(nums2)); // false
    }

}

File: contains-duplicate/solution217.py
# https://leetcode.com/problems/contains-duplicate/

from typing import List


class Solution217:

    # T: O(n), S: O(n)
    def containsDuplicate(self, nums: List[int]) -> bool:
        return len(set(nums)) != len(nums)

    # T: O(n log n), S: O(1)
    def containsDuplicateSort(self, nums: List[int]) -> bool:
        nums.sort()
        for i in range(1, len(nums)):
            if nums[i] == nums[i - 1]:
                return True
        return False


def main() -> None:
    sol = Solution217()
    nums1 = [1, 2, 3, 1]
    nums2 = [1, 2, 3, 4]
    print(sol.containsDuplicate(nums1))  # true
    print(sol.containsDuplicateSort(nums1))  # true
    print(sol.containsDuplicate(nums2))  # false
    print(sol.containsDuplicateSort(nums2))  # false


if __name__ == "__main__":
    main()
